converter/src/main/java/gov/cms/qpp/conversion/validate/CpcMeasureDataValidator.java

```diff
@@ -0,0 +1,158 @@
+	 * @param supplementalDataType current data type to be validated
+	 */
+	private void validateAllSupplementalNodesOfSpecifiedType(
+			Node node, TemplateId currSupplementalDataTemplateId, String supplementalDataType) {
+		Set<Node> supplementalDataNodes =
+				node.getChildNodes(currSupplementalDataTemplateId).collect(Collectors.toSet());
+		EnumSet<SupplementalData> codes = EnumSet.copyOf(
+				 SupplementalData.getSupplementalDataSetByType(supplementalDataType));
+
+		for (SupplementalData supplementalData : codes) {
+			Node validatedSupplementalNode = filterCorrectNode(
+					supplementalDataType, supplementalDataNodes, supplementalData);
+
+			if (validatedSupplementalNode == null) {
+				addSupplementalValidationError(node, supplementalData);
+			}
+		}
+		validateSupplementalDataNodeCounts(node, supplementalDataType, supplementalDataNodes);
+	}
+
+	/**
+	 * Filters out the specified Supplemental Data Node to validate
+	 *
+	 * @param supplementalDataType Supplemental value name for the current Node value
+	 * @param supplementalDataNodes List of nodes to filter
+	 * @param supplementalData Current Supplemental Data to validate against
+	 * @return
+	 */
+	private Node filterCorrectNode(
+			String supplementalDataType, Set<Node> supplementalDataNodes, SupplementalData supplementalData) {
+		return supplementalDataNodes.stream()
+				.filter(filterDataBySupplementalCode(supplementalDataType, supplementalData.getCode()))
+				.findFirst()
+				.orElse(null);
+	}
+
+	/**
+	 * Predicate for filtering the correct Supplemental Data node.
+	 *
+	 * @param nodeValueName Object that maps to the current node value
+	 * @param code Required code to be validate against
+	 * @return
+	 */
+	private Predicate<Node> filterDataBySupplementalCode(String nodeValueName, String code) {
+		return thisNode -> code.equalsIgnoreCase(thisNode.getValue(nodeValueName));
+	}
+
+	/**
+	 * Adds a Supplemental Validation Error upon failure
+	 *
+	 * @param node Object being validated
+	 * @param supplementalData Object holding the current code that was validated
+	 */
+	private void addSupplementalValidationError(Node node, SupplementalData supplementalData) {
+		MeasureConfig config =
+				MeasureConfigs.getConfigurationMap()
+						.get(node.getParent().getValue(QualityMeasureIdValidator.MEASURE_ID));
+		if (config != null) {
+			LocalizedError error =
+					ErrorCode.CPC_PLUS_MISSING_SUPPLEMENTAL_CODE.format(supplementalData.getCode(),
+							config.getElectronicMeasureId(), node.getValue(MeasureDataDecoder.MEASURE_TYPE));
+			addValidationError(Detail.forErrorAndNode(error, node));
+		}
+	}
+
+	/**
+	 * Validates all Supplemental Data nodes for an Aggregate count
+	 *
+	 * @param node parent node
+	 * @param supplementalDataType type of supplemental nodes
+	 * @param supplementalDataNodes supplemental nodes to be validated
+	 */
+	private void validateSupplementalDataNodeCounts(
+			Node node, String supplementalDataType, Set<Node> supplementalDataNodes) {
```

Comment:
The reason for this is because the line goes over the line length limit and I didn't want sonar to complain about it.

Reply:
Such is the problem with automated review tools. Too much structure makes for ugly workarounds :(

Reply:
Yeah I dislike it as well :(

Reply:
This will be fixed momentarily.